client/src/util/csvExport.ts

```diff
@@ -0,0 +1,192 @@
+  }
+  const stringValue = String(value);
+  if (
+    stringValue.includes(",") ||
+    stringValue.includes('"') ||
+    stringValue.includes("\n")
+  ) {
+    return `"${stringValue.replace(/"/g, '""')}"`;
+  }
+  return stringValue;
+}
+
+/**
+ * Converts an array of values to a CSV row
+ */
+function arrayToCsvRow(values: any[]): string {
+  return values.map(escapeCsvValue).join(",");
+}
+
+/**
+ * Exports categories and label annotations as CSV
+ */
+export async function exportCategoriesCSV(
+  annoMatrix: AnnoMatrix,
+  schema: Schema
+): Promise<void> {
+  try {
+    // Get all categorical columns from schema (exclude continuous metadata)
+    const allColumns = Object.keys(schema.annotations.obsByName);
+    const categoricalColumns = allColumns.filter(
+      (col) =>
+        !EXCLUDED_CATEGORY_NAMES.includes(col) &&
+        isCategoricalAnnotation(schema, col) === true
+    );
+
+    // Include name_0 (cell index) as the first column
+    const cellIndexColumn = schema.annotations.obs.index;
+
+    // Categorize columns, excluding the cell index column to prevent duplication
+    const standardColumns = categoricalColumns.filter(
+      (col) => STANDARD_CATEGORY_NAMES.includes(col) && col !== cellIndexColumn
+    );
+    const authorColumns = categoricalColumns.filter(
+      (col) =>
+        !STANDARD_CATEGORY_NAMES.includes(col) &&
+        !schema.annotations.obsByName[col]?.writable &&
+        col !== cellIndexColumn
+    );
+    const userColumns = categoricalColumns.filter(
+      (col) =>
+        schema.annotations.obsByName[col]?.writable && col !== cellIndexColumn
+    );
+
+    const allColumnsInOrder = [
+      ...standardColumns,
+      ...authorColumns,
+      ...userColumns,
+    ];
+
+    // Fetch all the data (excluding the cell index column as it's already available)
+    if (allColumnsInOrder.length > 0) {
+      await annoMatrix.fetch(Field.obs, allColumnsInOrder);
+    }
+    const obsData = annoMatrix._cache.obs;
+
+    // Create CSV content
+    const csvRows: string[] = [];
+
+    // Row 1: Block headers
+    const blockHeaders = [
+      "Cell Index",
+      ...standardColumns.map(() => "Standard Categories"),
+      ...authorColumns.map(() => "Author Categories"),
+      ...userColumns.map(() => "User Categories"),
+    ];
+    csvRows.push(arrayToCsvRow(blockHeaders));
+
+    // Row 2: Column names
+    const allColumnNames = [cellIndexColumn, ...allColumnsInOrder];
+    csvRows.push(arrayToCsvRow(allColumnNames));
+
+    // Data rows
+    for (let i = 0; i < annoMatrix.nObs; i += 1) {
+      const rowData = [cellIndexColumn, ...allColumnsInOrder].map((col) => {
+        if (col === cellIndexColumn) {
+          // For cell index, use the row label from rowIndex
+          return obsData.rowIndex.getLabel(i) || i.toString();
+        }
+        const column = obsData.col(col);
+        // Use getLabelArray to get the proper string labels instead of raw codes
+        return column.getLabelArray()[i];
```

Comment:
This seems horribly inefficient. getLabelArray returns a full array copied in-memory and [i] indexes out one element.

There are other functions in dataframe.ts that are designed to get the value/label at a particular index.